Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize == pageSize);
+
+            // Purge entries whose SlidingExpiration has passed.
+            do
+            {
+                QuerySnapshot querySnapshot = await
+                    _cacheEntries.OrderBy("LastRefresh")
+                    .Limit(pageSize)
+                    .GetSnapshotAsync(token);
+                batchSize = 0;
+                WriteBatch writeBatch = _cacheEntries.Database.StartBatch();
+                foreach (DocumentSnapshot docSnapshot in querySnapshot.Documents)
+                {
+                    CacheDoc doc = docSnapshot.ConvertTo<CacheDoc>();
+                    if (doc.SlidingExpirationSeconds.HasValue)
+                    {
+                        var slidingExpiration =
+                            doc.LastRefresh.GetValueOrDefault()
+                            + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.Value);
+                        if (slidingExpiration < now)
+                        {
+                            writeBatch.Delete(docSnapshot.Reference,
+                                Precondition.LastUpdated(
+                                    docSnapshot.UpdateTime.GetValueOrDefault()));
+                            batchSize += 1;
+                        }
+                    }
+                }
+                if (batchSize > 0)
+                {
+                    _logger.LogDebug("Collecting {0} cache entries.", batchSize);
+                    await writeBatch.CommitAsync(token);
+                }
+                if (token.IsCancellationRequested)
+                {
+                    return;
+                }
+            } while (batchSize > 0);
+            _logger.LogTrace("End garbage collection.");
+        }
+
+        internal static string GetProjectId()
+        {
+            // Use the service account credentials, if present.
+            GoogleCredential googleCredential = Google.Apis.Auth.OAuth2
+                .GoogleCredential.GetApplicationDefault();
+            if (googleCredential != null)
+            {
+                ICredential credential = googleCredential.UnderlyingCredential;
+                ServiceAccountCredential serviceAccountCredential =
+                    credential as ServiceAccountCredential;
+                if (serviceAccountCredential != null)
+                {
+                    return serviceAccountCredential.ProjectId;
+                }
+            }
+            try
+            {
+                // Query the metadata server.
+                HttpClient http = new HttpClient();
+                http.DefaultRequestHeaders.Add("Metadata-Flavor", "Google");
+                http.BaseAddress = new Uri(
+                    @"http://metadata.google.internal/computeMetadata/v1/project/");
+                return http.GetStringAsync("project-id").Result;

```

Comment:
Task.Run would be safer here, unless we're really sure of the context in which it's called.